wolfssl/src/ssl.rs

```diff
@@ -1100,6 +1111,95 @@ impl<IOCB: IOCallbacks> Session<IOCB> {
+
+        // SAFETY: No documentation available for [`wolfSSL_KeepArrays`] [`wolfSSL_set_tls13_secret_cb`].
+        // But based on api implementation, it expects a valid pointer to `WOLFSSL`.
+        //
+        // This implementation is taken from following examples:
+        // https://github.com/wolfSSL/wolfssl-examples/blob/master/tls/client-tls13.c
+        // https://github.com/wolfSSL/wolfssl-examples/blob/master/tls/server-tls13.c
+        match unsafe {
+            let ssl = self.ssl.lock().as_ptr();
+            wolfssl_sys::wolfSSL_KeepArrays(ssl);
+            wolfssl_sys::wolfSSL_set_tls13_secret_cb(
+                ssl,
+                Some(Self::tls13_secret_cb),
+                keylog_c as *mut c_void,
+            )
+        } {
+            wolfssl_sys::WOLFSSL_SUCCESS => Ok(()),
+            wolfssl_sys::WOLFSSL_FATAL_ERROR => panic!("No SSL context"),
+            e => unreachable!("{e:?}"),
+        }
+    }
+
+    #[cfg(feature = "debug")]
+    unsafe extern "C" fn tls13_secret_cb(
+        ssl: *mut wolfssl_sys::WOLFSSL,
+        id: ::std::os::raw::c_int,
+        secret: *const ::std::os::raw::c_uchar,
+        secret_size: ::std::os::raw::c_int,
+        ctx: *mut ::std::os::raw::c_void,
+    ) -> ::std::os::raw::c_int {
+        debug_assert!(!ssl.is_null());
+        debug_assert!(!secret.is_null());
+        debug_assert!(!ctx.is_null());
+
+        let keylog = Box::from_raw(ctx as *mut Arc<dyn Tls13SecretCallbacks>);
+
+        let mut secret_keylog = match id as wolfssl_sys::Tls13Secret {
+            wolfssl_sys::Tls13Secret_CLIENT_EARLY_TRAFFIC_SECRET => "CLIENT_EARLY_TRAFFIC_SECRET",
+            wolfssl_sys::Tls13Secret_CLIENT_HANDSHAKE_TRAFFIC_SECRET => {
+                "CLIENT_HANDSHAKE_TRAFFIC_SECRET"
+            }
+            wolfssl_sys::Tls13Secret_SERVER_HANDSHAKE_TRAFFIC_SECRET => {
+                "SERVER_HANDSHAKE_TRAFFIC_SECRET"
+            }
+            wolfssl_sys::Tls13Secret_CLIENT_TRAFFIC_SECRET => "CLIENT_TRAFFIC_SECRET_0",
+            wolfssl_sys::Tls13Secret_SERVER_TRAFFIC_SECRET => "SERVER_TRAFFIC_SECRET_0",
+            wolfssl_sys::Tls13Secret_EARLY_EXPORTER_SECRET => "EARLY_EXPORTER_SECRET",
+            wolfssl_sys::Tls13Secret_EXPORTER_SECRET => "EXPORTER_SECRET",
+            _e => "UNKNOWN_SECRET",
+        }
+        .to_string();
+
+        secret_keylog.push(' ');
+
+        let mut random: Vec<u8> = vec![0; RANDOM_SIZE];
+        let random_size = if keylog.is_server() {
+            wolfssl_sys::wolfSSL_get_server_random(
+                ssl,
+                random.as_mut_ptr() as *mut c_uchar,
+                random.len(),
+            )
+        } else {
+            wolfssl_sys::wolfSSL_get_client_random(
+                ssl,
+                random.as_mut_ptr() as *mut c_uchar,
+                random.len(),
+            )
+        };
+        (0..random_size).for_each(|i| secret_keylog.push_str(&format!("{:02x}", random[i])));
```

Comment:
This feels like a `for i in random.iter() { ...}` loop or a `random.iter().foreach()` in a funny hat to me.

But as with the enum above why are we stringifying this stuff at this level? It would seem more natural if `Tls13SecretCallbacks::secret` took `which: Tls13Secret, random: &[u8], secret: &[u8]` as arguments?

Reply:
Added two methods based on our discussion
One to get as a wireshark keyfile line
Another to get secret and random separately